framework/audio/midi/internal/sequencer.cpp

```diff
@@ -58,6 +49,53 @@ void Sequencer::init(float samplerate, float gain)
     });
 }
 
+void Sequencer::loadMIDI(const std::shared_ptr<MidiStream>& stream)
+{
+    m_midiStream = stream;
+    m_streamState = StreamState();
+
+    m_midiData = stream->initData;
+
+    m_midiStream->stream.onReceive(this, [this](const MidiData& data) { onDataReceived(data); });
+    m_midiStream->stream.onClose(this, [this]() { onStreamClosed(); });
+
+    requestData(0);
+
+    buildTempoMap();
+    synth()->loadSF(m_midiData.programs(), "", [this](uint16_t percent) {
+        LOGI() << "sf loading: " << percent;
+    });
+}
+
+void Sequencer::requestData(uint32_t tick)
+{
+    LOGI() << "requestData: " << tick;
+    if (m_streamState.closed) {
+        LOGE() << "stream closed";
+        m_streamState.requested = false;
+        return;
+    }
+    m_streamState.requested = true;
+    m_midiStream->request.send(tick);
+}
+
+void Sequencer::onDataReceived(const MidiData& data)
+{
+    LOGI() << "onDataReceived: " << data.tracks.front().channels.front().events.front().tick;
+    //! TODO implement merge
+    m_midiData = data;
+    m_streamState.requested = false;

```

Comment:
Can you explain a bit the relations between the midi data and the stream state?

I'm curious, why do we set "requested = false" here.

Reply:
midi data is data (midi events)
if there is no data to play, then we consider that we are finished.
But the data may not be there because they just haven't come to us yet.
To handle this situation, when requesting new data, we set the flag (requested = true), indicating that we have requested new data.
When new data received, we clear this flag (requested = false).
And when checking for the end, we check if the flag is set, then we should not return that we have ended, because we do not know this yet.